### PR TITLE
Add safe zero-copy conversion from bytes::Bytes (#4254)

### LINKED ISSUE #4254: Safe Zero-Copy Buffer from bytes::Bytes
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The Rust ecosystem makes extensive use of [bytes::Bytes](https://docs.rs/bytes/latest/bytes/struct.Bytes.html) to represent a shared slice of data. Providing safe zero-copy interoperability with this type will therefore allow applications to avoid unnecessary copying of potentially large IPC payloads

**Describe the solution you'd like**

Provide a `From<bytes::Bytes>` implementation for `Buffer`

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4254

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-flight/src/decode.rs [hunk 1]
```diff
@@ -258,7 +259,7 @@ impl FlightDataDecoder {
                     ));
                 };
 
-                let buffer: arrow_buffer::Buffer = data.data_body.into();
+                let buffer = Buffer::from_bytes(data.data_body.into());
```
Comment (tustvold) [inline arrow-flight/src/decode.rs:262]: This API is kind of unfortunate, but I've not been able to find a good way to workaround the blanket From<AsRef<[u8]>> impl for Buffer

File: arrow-buffer/Cargo.toml [hunk 2]
```diff
@@ -34,6 +34,7 @@ path = "src/lib.rs"
 bench = false
 
 [dependencies]
+bytes = { version = "1.4" }
```
Comment (tustvold) [inline arrow-buffer/Cargo.toml:37]: This is such a common dependency that I don't see this being overly burdensome

Comment (tustvold) [reply arrow-buffer/Cargo.toml:37]: Because the bytes crate doesn't allow foreign allocations

File: arrow-flight/src/decode.rs (hunk 1 above)
Comment (alamb) [reply arrow-flight/src/decode.rs:262]: 👍  I think we can use this API in IOx as well, FWIW.

**Review by crepererum — APPROVED** [review]

**Review by alamb — APPROVED** [review]
LGTM -- thank you @tustvold

File: arrow-buffer/Cargo.toml (hunk 2 above)
Comment (alamb) [reply arrow-buffer/Cargo.toml:37]: Perhaps we could file a request upstream?  I don't see anything related to this functionality in the currently open issues

https://github.com/tokio-rs/bytes/issues?q=is%3Aissue+is%3Aopen+allocation+

Comment (tustvold) [reply arrow-buffer/Cargo.toml:37]: https://github.com/tokio-rs/bytes/issues/437 is the upstream ticket

**Pull request merged**